File: Custom_sort_string.java
// Time Complexity : O(N)max of length of s,order
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
    public String customSortString(String order, String s) {
        HashMap<Character,Integer> map=new HashMap<>();
        for(int i=0;i<s.length();i++){
            char c=s.charAt(i);
            if(!map.containsKey(c)){
                map.put(c,0);
            }
            map.put(c,map.get(c)+1);
        }
        StringBuilder st=new StringBuilder();
        for(int i=0;i<order.length();i++){
            char c=order.charAt(i);
            if(map.containsKey(c)){
                int count=map.get(c);
                while(count>0){
                    st.append(c);
                    count--;
                }
                map.remove(c);
            }
        }
        
        for(char c:map.keySet()){
             int count=map.get(c);
                while(count>0){
                    st.append(c);
                    count--;
                }
        }
        
        return st.toString();
    }
}

File: CustomSortString.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Algorithms
{
    public class CustomSortString
    {
        /// Time Complexity : O(n) 
        // Space Complexity :O(1)
        // Did this code successfully run on Leetcode :Yes
        // Any problem you faced while coding this :  Referred Online
        public string CustomSortString_1(string order, string s)
        {
            Dictionary<char, int> map = new Dictionary<char, int>();
            for (int i = 0; i < s.ToCharArray().Length; i++)
            {
                if (map.ContainsKey(s.ToCharArray()[i]))
                {
                    map[s.ToCharArray()[i]] = map.GetValueOrDefault(s.ToCharArray()[i]) + 1;
                }
                else
                {
                    map.Add(s.ToCharArray()[i], 1);
                }
            }
            StringBuilder sb = new StringBuilder();
            for (int j = 0; j < order.ToCharArray().Length; j++)
            {
                if (map.ContainsKey(order.ToCharArray()[j]))
                {
                    int charCount = map.GetValueOrDefault(order.ToCharArray()[j]);
                    while (charCount > 0)
                    {
                        sb.Append(order.ToCharArray()[j]);
                        charCount--;
                    }
                    map.Remove(order.ToCharArray()[j]);
                }
            }

            foreach (var item in map)
            {
                int charCount = item.Value;
                while (charCount > 0)
                {
                    sb.Append(item.Key);
                    charCount--;
                }
            }
            return sb.ToString();
        }
    }
}
